prov/shm/src/smr_ep.c

```diff
@@ -1697,16 +1305,25 @@ static int smr_ep_ctrl(struct fid *fid, int command, void *arg)
 			domain = container_of(ep->util_ep.domain,
 					      struct smr_domain,
 					      util_domain.domain_fid);
-			ret = smr_ep_srx_context(domain, ep->rx_size,
-						 &ep->srx);
+			ret = util_ep_srx_context(&domain->util_domain,
+					ep->rx_size, SMR_IOV_LIMIT,
+					SMR_INJECT_SIZE, &ep->srx);
 			if (ret)
 				return ret;
-			ret = smr_srx_bind(&ep->srx->fid,
+
+			util_get_peer_srx(ep->srx)->peer_ops = &smr_srx_peer_ops;
+			ret = util_srx_bind(&ep->srx->fid,
 					   &ep->util_ep.rx_cq->cq_fid.fid,
 					   FI_RECV);
 			if (ret)
 				return ret;
 		} else {
+			srx = calloc(1, sizeof(*srx));
```

Comment:
I am not sure how this could work as peer provider get nothing but the owner_ops from the owner provider. If owner provider does not use `util_ep_srx_context` to initiate fid_peer_srx (as part of util_srx_ctx), how could peer provider access the resources (like queues, cq, etc.) in the owner provider when calling `get_msg/tag`? I see this change still make peer provider get the imported peer_srx's `ep_fid.fid.context`, which will be the `srx_ctx` in the owner provider. But I do not see the man page forces owner provider to setup `fid_peer_srx` resource in this way?

Reply:
@shijin-aws I'm not sure I understand your question. I have not updated the PR to include the fix which srx it uses when processing the unexpected message as I'm working on other updates to include with the next push as well.
What is the concern with initialization? What it does here is either allocate a util srx if it is not already bound to something (and it will create its own srx that it will use directly), or it creates a peer srx (without the resources since it will use the owner's) and imports the owner ops while setting its own peer ops. This is the srx that will be passed to the owner when getting a buffer/queuing a message.
Can you elaborate more on the concern?

Reply:
> or it creates a peer srx (without the resources since it will use the owner's)

Yes, this is my question, how does peer provider can access owner's resouces (in util_srx_ctx) from this fresh new `fid_peer_srx`? I guess you make it work by doing 

```
srx->ep_fid.fid.context =
				smr_get_peer_srx(ep)->ep_fid.fid.context;
```
So this new srx still has the context that was pointed to owner's `util_srx_ctx`.  Do I understand it correctly? My question is what if owner does not create its resource via `util_ep_srx_context` ?

Reply:
Oh I see your question. 
Yes, we get the context from the ep->srx which is a generic fid_peer_srx type (imported directly from the owner), so even if the owner is not using the util resources, it shouldn't matter. We will always just use the peer_srx->owner_ops.
The import of the srx is defines as follows by the man page:
`Peer SRXs are configured by the owner calling the peer’s fi_srx_context() call with the FI_PEER flag set. The context parameter passed to fi_srx_context() must be a struct fi_peer_srx_context..`
And the fi_peer_srx_context is defines as:
```
struct fid_peer_srx {
    struct fid_ep ep_fid;
    struct fi_ops_srx_owner *owner_ops;
    struct fi_ops_srx_peer *peer_ops;
};

struct fi_peer_srx_context {
    size_t size;
    struct fid_peer_srx *srx;
};
```
So when the peer imports fi_peer_srx_context, we save the entire fid_peer_srx *srx which is how we get to the owner ops

Reply:
I know I have bothered you on this many times so thank you for your patience in advance :)

The proposed workflow (I will just use shm as the peer)
1. Owner provider sets up it's own `struct fid_peer_srx` resource (in whatever way) as `owner_peer_srx`;
2. Owner provider sets up shm provider resources (info, domain, ep, etc.)
3. Owner provider creates `struct fi_peer_srx_context srx_context` and assign `srx_context.srx = owner_peer_srx`
4. Owner provider has an `attr.op_flags |= FI_PEER` and  call `fi_srx_context(shm_domain, attr, &dummy_srx, srx_context)`. In shm provider, `smr_srx_context` will set `smr_domain->srx = ((struct fi_peer_srx_context *) (context))->srx;` when there is a `FI_PEER` bit in `attr.op_flags`, and return success without creating a new `fid_peer_srx` object.
5. (This one is not listed in the example in man page but I think we need to do it according to code read and our earlier conversation) owner provider calls `fi_ep_bind(shm_ep_fid, owner_peer_srx.ep_fid, 0)`. In `smr_ep_bind`, it will finally set `ep->srx = owner_peer_srx`
6. Owner provider call fi_enable(shm_ep). `smr_ep_ctrl` will check if `ep->srx` is already set. And because of step 5 this is set. Then in this PR `smr_ep_ctrl` does 
```
srx = calloc(1, sizeof(*srx));
srx->peer_ops = &smr_srx_peer_ops;
			srx->owner_ops = smr_get_peer_srx(ep)->owner_ops;
			srx->ep_fid.fid.context =
				smr_get_peer_srx(ep)->ep_fid.fid.context;
			ep->srx = &srx->ep_fid;
```
which means it create a fresh new `fid_peer_srx`,  inherit  the owner ops from the owner provider as `srx->owner_ops = smr_get_peer_srx(ep)->owner_ops`, and the `ep_fid.fid.context` from the owner provider.

Now the issue is the owner_ops like `get_msg` need to access owner provider resources, e.g. packet pool, recv queue, in some way. And this "some way" really depends on how this `owner_peer_srx` was created in step 1. For example, in efa provider, I would like to make owner_peer_srx as a field in `efa_ep` struct, which has all the resources I need, so I will call some function like `efa_create_peer_srx(efa_ep, &efa_ep->peer_srx)` to create this `peer_srx` as part of an `efa_ep` object;  Then to implement efa provider's `get_msg` I will need to get `efa_ep` from the input `srx` variable. An easy and natural way to do that is to do this inside the `get_msg` call:
`efa_ep = container_of(srx, struct efa_ep, efa_ep.peer_srx)`.

But with your PR, this workflow will not work any more as the final `ep->srx` in shm provider will not be part of `efa_ep` because it is created by `calloc`. Then when shm_provider call `ep->srx->peer_ops->owner_ops->get_msg` it will likely to segfault because  `efa_ep = container_of(srx, struct efa_ep, efa_ep.peer_srx)` will not point to a valid `efa_ep` pointer.

In my mind this issue has 2 potential solutions:

Solution 1: efa provider needs to do `peer_srx.ep_fid.fid.context = efa_ep`, and retrieve `efa_ep` back in `get_msg` as `efa_ep = srx.ep_fid.fid.context` because shm provider's srx copied this information. 

Solution 2: Do not create a fresh new srx via calloc, but just make `ep->srx->peer_ops = &smr_srx_peer_ops;`, which means you just replace the peer_ops in the imported srx.

One final question is that I still do not see how `smr_domain->srx` was finally used by shm provider in the workflow I proposed above. If it is not used, what is the influence of Step 3 and 4 in my workflow?

I hope the explanation to my issue makes sense and It will be greatly appreciated if you could give me some feedback on them, thanks!

Reply:
@shijin-aws No worries, I'm happy to explain my vision as much as you need!
The util srx implementation sets the peer_srx context to be the util_srx
`srx->peer_srx.ep_fid.fid.context = srx;`
srx here is of type struct util_srx_ctx * which is what owns all of the resources (rx pool, queues, etc)
So when we set the duplicate srx (for the peer/shm), we set its context to be the same as the imported context:
`srx->ep_fid.fid.context = smr_get_peer_srx(ep)->ep_fid.fid.context;`
When we pass in the duplicate srx, it will have the same context and owner ops as the owner so we will call the owner's get_msg which will point to the util get_msg code which will cast the context of the srx into a util_srx (since it set it, it knows that it is a util_srx type)
`srx_ctx = srx->ep_fid.fid.context;`
If a provider is not using the util implementation, that is fine; it will just have to set the context to a pointer that will help it get to its resources. I would assume that this would be separate from the EP since the srx is not owned by a single EP. If you need a specific piece of information during get/queue that is currently in the EP, then I would think it would need to be moved to the srx set of resources instead.
Also I want to note that the srx fid is of type ep_fid because it can be passed in to a fi_recv call so I'm not sure when you say "get to the efa_ep", I'm not sure if you mean the efa_ep for an EP or the srx ep_fid.
I hope that makes sense?
In regards to your question about the domain - it's not necessarily that it is needed/used, it's just that when sharing the srx, the resource does not belong to a single EP so we associate it with a domain instead. As far as whether that is a separate domain per peer or a shared domain is up to interpretation (see the first conversation in this PR which discusses the shared domain question).

Reply:
Thanks a lot! I think your comment answers all of my questions

> If a provider is not using the util implementation, that is fine; it will just have to set the context to a pointer that will help it get to its resources. I would assume that this would be separate from the EP since the srx is not owned by a single EP. If you need a specific piece of information during get/queue that is currently in the EP, then I would think it would need to be moved to the srx set of resources instead.

Yes, I think shm choose to move these resources to `util_srx_ctx` because it wants to support the public `fi_srx_context` API which makes the srx can be shared among multiple EP. But if the owner provider is not going support `fi_srx_context` that allow user to access SRX and bind it to multiple EPs, I think it should be fine to make srx access resources in owner's ep directly.

> Also I want to note that the srx fid is of type ep_fid because it can be passed in to a fi_recv call so I'm not sure when you say "get to the efa_ep", I'm not sure if you mean the efa_ep for an EP or the srx ep_fid.

I was just making an example of how the `get_msg` ops can get to the owner resources from the input `fid_peer_srx` variable. It does not relate to the ep_fid in fi_recv. Now it's clear to me that I need to access the resource via `ep_fid.fid.context`. So this is not an issue.

> In regards to your question about the domain - it's not necessarily that it is needed/used, it's just that when sharing the srx, the resource does not belong to a single EP so we associate it with a domain instead. As far as whether that is a separate domain per peer or a shared domain is up to interpretation (see the first conversation in this PR which discusses the shared domain question).

I see. Just want to double check: If I only want to make `shm_ep->srx` get the owner_ops and context from owner's peer_srx, the `fi_ep_bind(shm_ep, ..., owner_peer_srx)` should be enough, right?

Reply:
It might be enough for your specific case but I think it makes the peer vs generic srx case foggy. The combination of calling fi_srx_context with the FI_PEER flag + fi_ep_bind makes it clear that the expectation is to go through the peer API. If we're just calling fi_ep_bind, it might not be explicit enough to fit the definition. I think that's why we defined the peer flow to go through both calls.

Reply:
I see, I will follow this flow in my implementation, thanks!